### PR TITLE
adapter: Refactor ConnectionId to be an owned handle

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR fixes a known issue where we free `ConnectionId`s too early, which can possibly lead to their re-use and a subsequent panic because two different `Session`s within the adapter end up having the same ID. Currently we free a `ConnectionId` when a `ConnClient` `Drop`s, the issue with this is other parts of the system (especially since we've started moving operations of the coordinator thread) still have the now freed `ConnectionId` and expect it to be alive/valid.

The fix is to change `ConnectionId` from being a type definition of `u32`, to being an owned reference counted "handle". Users can `Clone` a `ConnectionId` which increments the reference count, and only once all handles for that ID have been dropped, e.g. all operations using it have completed, do we free the id.

### Motivation

Fixes: #18996 

### Tips for reviewer

I broke this up into two commits, reviewing them separately might be a bit easier:

1. The implementation of `IdHandle<T>`, which is part of `mz_ore`.
2. Defines `type ConnectionId = IdHandle<u32>` and updates all callers.

One "oddity" to note is that we implement `Borrow<T> for IdHandle<T>`. Implementing `Borrow` isn't super common, which is why I wanted to call it out. But we do it such that you can use the inner type `T` to lookup an `IdHandle<T>` in a `BTreeMap`. For example, the pgwire protocol allows canceling a request by `connection_id` which is a `u32`. By implementing `Borrow` we're able to lookup the `IdHandle` with just this `u32`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a, internal bug fix

### CODE REVIEW

**Review by jkosh44 — APPROVED** [review]
Nice!

File: src/ore/src/id_gen.rs [hunk 1]
```diff
@@ -97,8 +108,155 @@ where
     }
 }
 
+/// A clone-able owned reference to an ID.
+///
+/// Once all of the [`IdHandle`]s referencing an ID have been dropped, we will then free the ID
+/// for later re-use.
+#[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub enum IdHandle<T: From<u8> + AddAssign + PartialOrd + Copy> {
```
Comment (mjibson) [inline src/ore/src/id_gen.rs:116]: I don't see where From<u8> or AddAssign are required. Are they needed or do I miss something?

File: src/ore/src/id_gen.rs [hunk 2]
```diff
@@ -97,8 +108,155 @@ where
...
+        fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+            self.id.partial_cmp(&other.id)
+        }
+    }
+
+    impl<T> Ord for IdHandleInner<T>
+    where
+        T: Ord + From<u8> + AddAssign + PartialOrd + Copy,
+    {
+        fn cmp(&self, other: &Self) -> std::cmp::Ordering {
```
Comment (mjibson) [inline src/ore/src/id_gen.rs:240]: For my learning: my read of the rust docs about the PartialOrd trait is that it doesn't implement the `cmp` fn. Where is it coming from? Also why do we need this?

**Review by mjibson — APPROVED** [review]

File: src/ore/src/id_gen.rs [hunk 3]
```diff
@@ -97,8 +108,155 @@ where
...
+    /// Creates a new [`IdHandle`] with the provided ID.
+    ///
+    /// Note: It is __entirely__ up to the caller to make sure the provided ID is not already being
+    /// used.
+    pub const fn new_static(id: T) -> Self {
+        IdHandle::Static(id)
+    }
+
+    /// Returns the ID associated with this [`IdHandle`].
+    pub fn val(&self) -> T {
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/ore/src/id_gen.rs:136]: I like the idea, but since in this case `T` is `Copy`, it would be fairly trivial for a user to get a `T` anyways. The current places we use this function we need a `T`, so I don't think in this case returning a `&T` would get us much

File: src/adapter/src/client.rs [hunk 4]
```diff
@@ -80,7 +85,7 @@ impl Handle {
 pub struct Client {
     build_info: &'static BuildInfo,
     inner_cmd_tx: mpsc::UnboundedSender<Command>,
-    id_alloc: Arc<IdAllocator<ConnectionId>>,
+    id_alloc: IdAllocator<ConnectionIdType>,
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/adapter/src/client.rs:88]: Exactly!

File: src/ore/src/id_gen.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/ore/src/id_gen.rs:116]: Unfortunately these are required on the enum definition because of the custom `Drop` impl. `fn free(...)` is only implemented on `IdAllocator` when `T` has these bounds. So we need the `Drop` impl to require these bounds for `T`. But Rust doesn't allow `Drop` to be implemented for only _some_ instances of a type, which would be the case if we required these bounds on `Drop`, but not on `IdHandle` itself.

tl;dr we need to carry these bounds for the entire type so we can use them in `Drop`.

File: src/ore/src/id_gen.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/ore/src/id_gen.rs:240]: This is the `Ord` trait which has `fn cmp` as a required method. `PartialOrd` is implemented directly above. We need this because `BTreeMap` requires that keys implement `Ord`, and `Ord` requires we implement `Eq` and `PartialOrd`. We manually implement these as opposed to deriving them because `IdAllocator` doesn't implement them, nor would we want the `IdAllocator` to be considered when comparing `IdHandle`s, it should be based on the id itself.

File: src/adapter/src/client.rs [hunk 5]
```diff
@@ -389,7 +393,7 @@ impl SessionClient {
 
     /// Cancels the query currently running on another connection.
     pub fn cancel_request(&mut self, conn_id: ConnectionId, secret_key: u32) {
-        self.inner_mut().cancel_request(conn_id, secret_key)
+        self.inner_mut().cancel_request(conn_id.val(), secret_key)
```
Comment (def-) [inline src/adapter/src/client.rs:396]: Apparently we have no cancellation test currently?

**Review by def- — COMMENTED** [review]
Checked coverage report, seems mostly tested: https://buildkite.com/materialize/coverage/builds/119

File: src/adapter/src/client.rs (hunk 5 above)
Comment (benesch) [reply src/adapter/src/client.rs:396]: We definitely do! Is it not getting picked up by coverage for some reason?

Comment (benesch) [reply src/adapter/src/client.rs:396]: https://github.com/MaterializeInc/materialize/blob/613026cac4c426fe0a78528680228f991a2c20ec/src/environmentd/tests/server.rs#L329-L331

**Comment by benesch** [comment]
Found a spare second to review this. Looks really good, but I had a few comments that I wrote up in PR form: https://github.com/MaterializeInc/materialize/pull/19541

File: src/adapter/src/client.rs (hunk 5 above)
Comment (benesch) [reply src/adapter/src/client.rs:396]: Ah! Indeed. That outer function is dead code. The inner function _is_ properly tested. Good spot.

I put up https://github.com/MaterializeInc/materialize/pull/19587 to remove the dead code.
